frontend/src/metabase/public/containers/PublicQuestion/PublicQuestion.tsx

```diff
@@ -0,0 +1,238 @@
+
+  const [card, setCard] = useState<Card | null>(null);
+  const [result, setResult] = useState<Dataset | null>(null);
+  const [parameterValues, setParameterValues] = useState<ParameterValuesMap>(
+    {},
+  );
+
+  useMount(async () => {
+    if (uuid) {
+      setPublicQuestionEndpoints(uuid);
+    } else if (token) {
+      setEmbedQuestionEndpoints(token);
+    }
+
+    try {
+      let card;
+      if (token) {
+        card = await EmbedApi.card({ token });
+      } else if (uuid) {
+        card = await PublicApi.card({ uuid });
+      } else {
+        throw { status: 404 };
+      }
+
+      if (card.param_values) {
+        await dispatch(addParamValues(card.param_values));
+      }
+      if (card.param_fields) {
+        await dispatch(addFields(card.param_fields));
+      }
+
+      const parameters = getCardUiParameters(
+        card,
+        metadata,
+        {},
+        card.parameters || undefined,
+      );
+      const parameterValuesById = getParameterValuesByIdFromQueryParams(
+        parameters,
+        location.query,
+      );
+
+      setCard(card);
+      setParameterValues(parameterValuesById);
+    } catch (error) {
+      console.error("error", error);
+      dispatch(setErrorPage(error));
+    }
+  });
+
+  const setParameterValue = async (parameterId: ParameterId, value: any) => {
+    setParameterValues(prevParameterValues => ({
+      ...prevParameterValues,
+      [parameterId]: value,
+    }));
+  };
+
+  const setParameterValueToDefault = (parameterId: ParameterId) => {
+    const parameters = getParameters();
+    const parameter = parameters.find(({ id }) => id === parameterId);
+    if (parameter) {
+      setParameterValue(parameterId, parameter.default);
+    }
+  };
+
+  const run = useCallback(async () => {
+    if (!card) {
+      return;
+    }
+
+    const parameters = card.parameters || getParametersFromCard(card);
+
+    try {
+      setResult(null);
+
+      let newResult;
+      if (token) {
+        // embeds apply parameter values server-side
+        newResult = await maybeUsePivotEndpoint(
+          EmbedApi.cardQuery,
+          card,
+        )({
+          token,
+          ...getParameterValuesBySlug(parameters, parameterValues),
+        });
+      } else if (uuid) {
+        // public links currently apply parameters client-side
+        const datasetQuery = applyParameters(card, parameters, parameterValues);
+        newResult = await maybeUsePivotEndpoint(
+          PublicApi.cardQuery,
+          card,
+        )({
+          uuid,
+          parameters: JSON.stringify(datasetQuery.parameters),
+        });
+      } else {
+        throw { status: 404 };
+      }
+
+      setResult(newResult);
+    } catch (error) {
+      console.error("error", error);
+      dispatch(setErrorPage(error));
+    }
+  }, [card, dispatch, parameterValues, token, uuid]);
+
+  useEffect(() => {
+    run();
+  }, [run]);
+
+  const getParameters = () => {
+    if (!card) {
```

Comment:
I couldn't verify the result by removing `|| !initialized` here. Leaving is as is makes sense if you look at the change [introduced here](https://github.com/metabase/metabase/pull/38395/files#diff-13411b02ed09ca24e298ae823ffee87579e38d2fed27f914d4c3d19ae6b6826aR181).

In short, we used to send `[]` when it's not `initialized` or doesn't have `results` yet. But I don't know what would happen if we send parameters when the `results` aren't there yet.

I'd leave the `|| !initialized` here just to be safe